Kernel/FourVelocity.m

```diff
@@ -9,55 +9,86 @@
+qr, qz, \[Lambda]local ,qr0, qz0, rprime, zprime, \[CapitalDelta], \[CapitalSigma], \[Omega], utContra,urContra,u\[Theta]Contra,uzContra,u\[Phi]Contra, utCo, urCo, u\[Theta]Co, u\[Phi]Co},
+
+(*Constants of Motion*)
+{En,L,Q}= {"\[ScriptCapitalE]","\[ScriptCapitalL]","\[ScriptCapitalQ]"}/.KerrGeoConstantsOfMotion[0,p,0,x];
+
+\[CapitalUpsilon]z = p/Sqrt[-3+p];
+
+{qr0,qz0} = initPhases;
+
+qz[\[Lambda]_] := \[Lambda] \[CapitalUpsilon]z + qz0;
+
+ 
+
+utContra= Function[{Global`\[Lambda]},Evaluate[Sqrt[p/(-3+p)] ], Listable];  
+urContra:= Function[{Global`\[Lambda]},Evaluate[0],Listable];
+u\[Theta]Contra = Function[{Global`\[Lambda]}, Evaluate[(Sqrt[((1-x^2)/(-3+p))] Sin[qz[Global`\[Lambda]]] )/(p Sqrt[1+(-1+x^2) Cos[qz[Global`\[Lambda]]]^2])],Listable];  
+u\[Phi]Contra = Function[{Global`\[Lambda]},Evaluate[x/(Sqrt[-3+p] (p+p (-1+x^2) Cos[qz[Global`\[Lambda]]]^2))],Listable]; 
+
+utCo =  Function[{Global`\[Lambda]},Evaluate[-En], Listable];
+urCo= Function[{Global`\[Lambda]},Evaluate[0],Listable];
+u\[Theta]Co=  Function[{Global`\[Lambda]},Evaluate[(p Sqrt[(1-x^2)/(-3+p)] Sin[qz[Global`\[Lambda]]])/ Sqrt[1+(-1+x^2) Cos[qz[Global`\[Lambda]]]^2]],Listable];   
+u\[Phi]Co= Function[{Global`\[Lambda]},Evaluate[L],Listable];
 
+<|"\!\(\*SuperscriptBox[\(u\), \(t\)]\)"->utContra, "\!\(\*SuperscriptBox[\(u\), \(r\)]\)"->urContra, "\!\(\*SuperscriptBox[\(u\), \(\[Theta]\)]\)"-> u\[Theta]Contra, "\!\(\*SuperscriptBox[\(u\), \(\[Phi]\)]\)"->   u\[Phi]Contra,
+"\!\(\*SubscriptBox[\(u\), \(t\)]\)"->utCo, "\!\(\*SubscriptBox[\(u\), \(r\)]\)"->urCo, "\!\(\*SubscriptBox[\(u\), \(\[Theta]\)]\)"-> u\[Theta]Co, "\!\(\*SubscriptBox[\(u\), \(\[Phi]\)]\)"->   u\[Phi]Co|>
 
-(* ::Text:: *)
-(*FixMe: Circular Equatorial Retrograde orbits don't normalize to -1*)
 
 
-KerrGeoVelocityMino[a_,p_,e_,x_,initPhases_,index_ ]:= Module[{En,L,Q,r,z,r1,r2,r3,r4,kr,zp,zm,kz, \[CapitalUpsilon]r, \[CapitalUpsilon]z, 
+] 
+
+
+(* ::Section:: *)
+(*Kerr*)
+
+
+(* ::Subsection::Closed:: *)
+(*Generic (Mino)*)
+
+
+KerrGeoVelocityMino[a_,p_,e_,x_,initPhases_]:= Module[{En,L,Q,r,z,r1,r2,r3,r4,kr,zp,zm,kz, \[CapitalUpsilon]r, \[CapitalUpsilon]z, 
 qr, qz, \[Lambda]local ,qr0, qz0, rprime, zprime, \[CapitalDelta], \[CapitalSigma], \[Omega], utContra,urContra,u\[Theta]Contra,uzContra,u\[Phi]Contra, utCo, urCo, u\[Theta]Co, u\[Phi]Co},
 
 (*Constants of Motion*)
 {En,L,Q}= {"\[ScriptCapitalE]","\[ScriptCapitalL]","\[ScriptCapitalQ]"}/.KerrGeoConstantsOfMotion[a,p,e,x];
 
-(*Roots*)
-r1 = p/(1-e);
-r2 = p/(1+e);
-zm = Sqrt[1-x^2];
+{r1,r2,r3,r4}=KerrGeodesics`OrbitalFrequencies`Private`KerrGeoRadialRoots[a, p, e, x,En,Q];
 
-(*Other Roots*)
-r3 = 1/(1-En^2) - (r1 + r2)/2 + Sqrt[(-(1/(1-En^2)) + (r1 + r2)/2 )^2 - (a^2 Q)/(r1 r2 (1 - En^2))];
-r4  =  (a^2  Q)/(r1 r2 r3 (1-En^2));
+{zp,zm}= KerrGeodesics`OrbitalFrequencies`Private`KerrGeoPolarRoots[a, p, e, x];
 
-zp = Sqrt[a^2 (1 - En^2) + (L^2)/(1 - zm^2) ]; 
 kr = ((r1-r2)(r3-r4))/((r1-r3)(r2-r4));
 
 kz = a^2 (1-En^2) zm^2/zp^2;
```

Comment:
Can these be removed now? These are the lines that lead to divide-by-zero at the separatrix.

Reply:
Do you mean the expressions for kr and kz?

These aren't singular at the separatrix. kz is finite and kr goes to 1. The problem is that EllipticK[kr] returns complex infinity when kr = 1 and that shows up in the frequencies, the analytic solutions for the coordinates and the four-velocities, with no obvious way (to me at least) for taking a finite limit exactly on the separatrix for generic orbits.

Reply:
When there is a double root, you get special solutions. These are currently not explicitly implemented in the package (except circular orbits). We should add a check for r2==r3 and produce an error message.

Reply:
Yes, I'm suggesting to remove the definitions of kr and kz. The problem with these is that in the case of repeated roots (eg. at the ISCO) they involve 0/0 and this causes problems. Since kr and kz are no longer used in this function, I think we can remove them to avoid warning messages. The special handling of repeated roots can then be done in KerrGeoMinoFrequencyr.